src/compiler/parser.ts

```diff
@@ -4527,6 +4527,21 @@ namespace ts {
                 }
             }
 
+            // Given:
+            //     x ? y => ({ y }) : z => ({ z })
+            // We try to parse the body of the first arrow function by looking at:
+            //     ({ y }) : z => ({ z })
+            // This is a valid arrow function with "z" as the return type.
+            //
+            // But, if we're in the true side of a conditional expression, this colon
+            // terminates the true side, so cannot be the return type. If we are in the
+            // false side, we may still be within the true side of a parent conditional
+            // expression, so don't allow it to be a return type either.
+            if (parsingConditionalExpression && token() === SyntaxKind.ColonToken) {
+                Debug.assert(!allowAmbiguity);

```

Comment:
can you issue an error in the `allowAmbiguity` case? At least I think that's what @DanielRosenwasser was proposing.

Edit: I'm not sure an error here will be (1) always correct (2) sensible even if correct. If it's not both, then making this happen is probably difficult.

Reply:
It can never be the case that we're parsing a conditional and allowing ambiguity, as the conditional only `tryParse`'s these to see if they are, then accepts them, otherwise tries some other expression type, hence the assert.